{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 20200422: Downstream analyses of 3D genomic changes in _L. pneumophila_ infected _A. castellanii_\n",
    "**cmdoret**\n",
    "\n",
    "In this notebook, I investigate the biological role of regions detected as changing in 3D structure during infection of amoeba by _Legionella pneumophila_. To summarise the context, this is a follow-up on 2 analyses:\n",
    "\n",
    "1. Genome assembly and annotation of _A. castellanii_ using ONT long reads, Hi-C, Illumina shotgun and RNAseq\n",
    "2. Pattern detection in Hi-C maps from _A. castellanii_ and quantification of their change during infection.\n",
    "\n",
    "I use the annotations generated in 1. to investigate potential roles of affected regions in 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load files and packages\n",
    "from typing import List, Union, Iterable, Optional\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import warnings\n",
    "from matplotlib import pyplot as plt\n",
    "import matplotlib.image as mpimg\n",
    "import hicstuff.hicstuff as hcs\n",
    "import hicstuff.view as hcv\n",
    "import cooler\n",
    "import pareidolia.io as pai\n",
    "import pareidolia.hic_utils as pah\n",
    "from cooltools.sample import sample_cooler\n",
    "\n",
    "\n",
    "out = '../../data/output/'\n",
    "indir = '../../data/input/'\n",
    "res = 2000\n",
    "samples = pd.read_csv('../../samples.tsv', sep='\\t', comment='#')\n",
    "infected_bedgraph = pd.read_csv(out + 'all_signals_AT420.bedgraph', sep='\\t')\n",
    "healthy_bedgraph = pd.read_csv(out + 'all_signals_AT421.bedgraph', sep='\\t')\n",
    "diff_borders = pd.read_csv(out + 'pareidolia/borders_change_infection_time.tsv', sep='\\t')\n",
    "diff_loops = pd.read_csv(out + 'pareidolia/loops_change_infection_time.tsv', sep='\\t')\n",
    "genes = pd.read_csv(indir + 'annotations/c3_annotations/Acanthamoeba_castellanii.annotations.txt', sep='\\t')\n",
    "genes = genes.rename(columns={\"Contig\": \"Chromosome\", \"Stop\": \"End\"})\n",
    "# Remove genes which do not have an entrez id\n",
    "#genes = genes.loc[~np.isnan(genes.entrezgene_id), :]\n",
    "#genes.entrezgene_id = genes.entrezgene_id.astype(int)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "      <th>Product</th>\n",
       "      <th>BUSCO</th>\n",
       "      <th>PFAM</th>\n",
       "      <th>InterPro</th>\n",
       "      <th>EggNog</th>\n",
       "      <th>COG</th>\n",
       "      <th>GO Terms</th>\n",
       "      <th>Secreted</th>\n",
       "      <th>Membrane</th>\n",
       "      <th>Protease</th>\n",
       "      <th>CAZyme</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Translation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FUN_000001</td>\n",
       "      <td>CDS</td>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>503</td>\n",
       "      <td>5457</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF00071</td>\n",
       "      <td>IPR001806;IPR020849;IPR027417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0016020;GO:0005525;GO:0003924;GO:0007165</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 (o833-852i)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MATKDFKIAIGGPGGVGKSAIVLQFVTGNYVSEYGAYRKQFALGDN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FUN_000002</td>\n",
       "      <td>CDS</td>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>7855</td>\n",
       "      <td>9815</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MQLMSTQHHKKRTPEDHRPDLPSSARNLLASSPPARSPRSPDGSPH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FUN_000003</td>\n",
       "      <td>CDS</td>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>10018</td>\n",
       "      <td>10602</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF07258</td>\n",
       "      <td>IPR017920</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MWRRPQSVAVGYTGAPKRASLAVGGSGPAVSGRKGDVRFPTLADLR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FUN_000004</td>\n",
       "      <td>CDS</td>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>10748</td>\n",
       "      <td>11427</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 (o90-111i)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MWKQSSRVAVVARSRAVVPSSSGLRRSAGALLPRRAGAETEPEWKL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FUN_000005</td>\n",
       "      <td>CDS</td>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>11709</td>\n",
       "      <td>30538</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF00076;PF00622;PF00632</td>\n",
       "      <td>IPR000504;IPR000569;IPR001870;IPR003877;IPR012...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0004842;GO:0003676;GO:0005515</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2 (o2988-3009i3316-3335o)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGAEEAVHLEGHHSSRSNRPATPQYSAASASGAAPPAPPLRGEAAS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       GeneID Feature  Chromosome  Start    End Strand Name  \\\n",
       "0  FUN_000001     CDS  scaffold_1    503   5457      -  NaN   \n",
       "1  FUN_000002     CDS  scaffold_1   7855   9815      +  NaN   \n",
       "2  FUN_000003     CDS  scaffold_1  10018  10602      -  NaN   \n",
       "3  FUN_000004     CDS  scaffold_1  10748  11427      +  NaN   \n",
       "4  FUN_000005     CDS  scaffold_1  11709  30538      -  NaN   \n",
       "\n",
       "                Product BUSCO                     PFAM  \\\n",
       "0  hypothetical protein   NaN                  PF00071   \n",
       "1  hypothetical protein   NaN                      NaN   \n",
       "2  hypothetical protein   NaN                  PF07258   \n",
       "3  hypothetical protein   NaN                      NaN   \n",
       "4  hypothetical protein   NaN  PF00076;PF00622;PF00632   \n",
       "\n",
       "                                            InterPro  EggNog  COG  \\\n",
       "0                      IPR001806;IPR020849;IPR027417     NaN  NaN   \n",
       "1                                                NaN     NaN  NaN   \n",
       "2                                          IPR017920     NaN  NaN   \n",
       "3                                                NaN     NaN  NaN   \n",
       "4  IPR000504;IPR000569;IPR001870;IPR003877;IPR012...     NaN  NaN   \n",
       "\n",
       "                                      GO Terms  Secreted  \\\n",
       "0  GO:0016020;GO:0005525;GO:0003924;GO:0007165       NaN   \n",
       "1                                          NaN       NaN   \n",
       "2                                          NaN       NaN   \n",
       "3                                          NaN       NaN   \n",
       "4             GO:0004842;GO:0003676;GO:0005515       NaN   \n",
       "\n",
       "                    Membrane Protease CAZyme  Notes  \\\n",
       "0              1 (o833-852i)      NaN    NaN    NaN   \n",
       "1                        NaN      NaN    NaN    NaN   \n",
       "2                        NaN      NaN    NaN    NaN   \n",
       "3               1 (o90-111i)      NaN    NaN    NaN   \n",
       "4  2 (o2988-3009i3316-3335o)      NaN    NaN    NaN   \n",
       "\n",
       "                                         Translation  \n",
       "0  MATKDFKIAIGGPGGVGKSAIVLQFVTGNYVSEYGAYRKQFALGDN...  \n",
       "1  MQLMSTQHHKKRTPEDHRPDLPSSARNLLASSPPARSPRSPDGSPH...  \n",
       "2  MWRRPQSVAVGYTGAPKRASLAVGGSGPAVSGRKGDVRFPTLADLR...  \n",
       "3  MWKQSSRVAVVARSRAVVPSSSGLRRSAGALLPRRAGAETEPEWKL...  \n",
       "4  MGAEEAVHLEGHHSSRSNRPATPQYSAASASGAAPPAPPLRGEAAS...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO :: Merging:\n",
      "../../data/output//cool/AT408.mcool::/resolutions/2000\n",
      "../../data/output//cool/AT419.mcool::/resolutions/2000\n",
      "../../data/output//cool/AT421.mcool::/resolutions/2000\n",
      "INFO :: NumExpr defaulting to 8 threads.\n",
      "INFO :: Creating cooler at \"healthy.cool::/\"\n",
      "INFO :: Writing chroms\n",
      "INFO :: Writing bins\n",
      "INFO :: Writing pixels\n",
      "INFO :: nnzs: [4627584, 552433, 1126529]\n",
      "INFO :: current: [4627584, 552433, 1126529]\n",
      "INFO :: Writing indexes\n",
      "INFO :: Writing info\n",
      "INFO :: Done\n",
      "INFO :: Merging:\n",
      "../../data/output//cool/PM106.mcool::/resolutions/2000\n",
      "../../data/output//cool/AT407.mcool::/resolutions/2000\n",
      "../../data/output//cool/AT418.mcool::/resolutions/2000\n",
      "../../data/output//cool/AT420.mcool::/resolutions/2000\n",
      "INFO :: Creating cooler at \"infected.cool::/\"\n",
      "INFO :: Writing chroms\n",
      "INFO :: Writing bins\n",
      "INFO :: Writing pixels\n",
      "INFO :: nnzs: [5264991, 4340619, 2188053, 1290478]\n",
      "INFO :: current: [5264991, 4340619, 2188053, 1290478]\n",
      "INFO :: Writing indexes\n",
      "INFO :: Writing info\n",
      "INFO :: Done\n"
     ]
    }
   ],
   "source": [
    "# Manage cools, merge samples by condition to increase coverage in visualisations\n",
    "def gather_and_merge_cools(condition: str, merged_out: str):\n",
    "    cools = (\n",
    "        samples.loc[samples.condition == condition, 'library']\n",
    "        .apply(lambda p: f\"{out}/cool/{p}.mcool::/resolutions/{res}\")\n",
    "        .tolist()\n",
    "    )\n",
    "    cooler.merge_coolers(merged_out, cools, mergebuf=10e8)\n",
    "    \n",
    "gather_and_merge_cools('uninfected', \"healthy.cool\")\n",
    "gather_and_merge_cools('infected', \"infected.cool\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO :: Creating cooler at \"sub_healthy.cool::/\"\n",
      "INFO :: Writing chroms\n",
      "INFO :: Writing bins\n",
      "INFO :: Writing pixels\n",
      "INFO :: Writing indexes\n",
      "INFO :: Writing info\n",
      "INFO :: Done\n",
      "INFO :: Creating cooler at \"sub_infected.cool::/\"\n",
      "INFO :: Writing chroms\n",
      "INFO :: Writing bins\n",
      "INFO :: Writing pixels\n",
      "INFO :: Writing indexes\n",
      "INFO :: Writing info\n",
      "INFO :: Done\n",
      "INFO :: variance is 207516.4035251037\n",
      "INFO :: variance is 178502.21226294566\n",
      "INFO :: variance is 6544.088298908443\n",
      "INFO :: variance is 22429.3587515201\n",
      "INFO :: variance is 1312.8879459398481\n",
      "INFO :: variance is 4614.821651772319\n",
      "INFO :: variance is 551.4385040972979\n",
      "INFO :: variance is 1275.4581852006963\n",
      "INFO :: variance is 272.55604743634456\n",
      "INFO :: variance is 416.4566990524887\n",
      "INFO :: variance is 131.60184953660084\n",
      "INFO :: variance is 149.6257068751345\n",
      "INFO :: variance is 60.93831076465593\n",
      "INFO :: variance is 56.90446190648478\n",
      "INFO :: variance is 27.298625543543995\n",
      "INFO :: variance is 22.41011815482649\n",
      "INFO :: variance is 11.949339108949578\n",
      "INFO :: variance is 9.020618365983388\n",
      "INFO :: variance is 5.149825973235855\n",
      "INFO :: variance is 3.6817146591182626\n",
      "INFO :: variance is 2.196722330977017\n",
      "INFO :: variance is 1.5160429161311808\n",
      "INFO :: variance is 0.9307435782830956\n",
      "INFO :: variance is 0.627830255579549\n",
      "INFO :: variance is 0.3926224162980588\n",
      "INFO :: variance is 0.2609529352746319\n",
      "INFO :: variance is 0.16514978354959262\n",
      "INFO :: variance is 0.10871929811909928\n",
      "INFO :: variance is 0.06933852140668614\n",
      "INFO :: variance is 0.04536409866690736\n",
      "INFO :: variance is 0.029076913633892334\n",
      "INFO :: variance is 0.01894719622124823\n",
      "INFO :: variance is 0.012183814727586\n",
      "INFO :: variance is 0.007918693444756432\n",
      "INFO :: variance is 0.0051026865999312795\n",
      "INFO :: variance is 0.0033108571096427107\n",
      "INFO :: variance is 0.00213635040887536\n",
      "INFO :: variance is 0.0013846583478133903\n",
      "INFO :: variance is 0.0008942401431206222\n",
      "INFO :: variance is 0.0005791877330755936\n",
      "INFO :: variance is 0.00037426253061453955\n",
      "INFO :: variance is 0.0002422949028631446\n",
      "INFO :: variance is 0.0001566246253751694\n",
      "INFO :: variance is 0.00010136788076730418\n",
      "INFO :: variance is 6.554187088967835e-05\n",
      "INFO :: variance is 4.241081427040553e-05\n",
      "INFO :: variance is 2.742593914889954e-05\n",
      "INFO :: variance is 1.7744586714406437e-05\n",
      "INFO :: variance is 1.147608498532111e-05\n",
      "INFO :: variance is 7.424438192813567e-06\n",
      "INFO :: variance is 347303.0609518247\n",
      "INFO :: variance is 56468.48096582146\n",
      "INFO :: variance is 9891.71681333544\n",
      "INFO :: variance is 4288.333938390718\n",
      "INFO :: variance is 1017.6090354500234\n",
      "INFO :: variance is 477.8640880629092\n",
      "INFO :: variance is 144.8791055175218\n",
      "INFO :: variance is 69.43952536288951\n",
      "INFO :: variance is 25.429483111342932\n",
      "INFO :: variance is 12.069046151152712\n",
      "INFO :: variance is 4.9579186690855\n",
      "INFO :: variance is 2.3218393387822323\n",
      "INFO :: variance is 1.0124807705183223\n",
      "INFO :: variance is 0.4700051397407853\n",
      "INFO :: variance is 0.21094010471692354\n",
      "INFO :: variance is 0.0974718424726765\n",
      "INFO :: variance is 0.04434223811512046\n",
      "INFO :: variance is 0.020446748274502456\n",
      "INFO :: variance is 0.009361404135577936\n",
      "INFO :: variance is 0.004313076181592695\n",
      "INFO :: variance is 0.0019808342209006245\n",
      "INFO :: variance is 0.0009124174357314829\n",
      "INFO :: variance is 0.0004196951377753369\n",
      "INFO :: variance is 0.00019332673617012742\n",
      "INFO :: variance is 8.90017563240869e-05\n",
      "INFO :: variance is 4.100274724800335e-05\n",
      "INFO :: variance is 1.888575521730221e-05\n",
      "INFO :: variance is 8.701970925408796e-06\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Subsample cools to identical coverage between conditions\n",
    "def subsample_cools(cools: List[str], cpus: int=8, balance: bool=True) -> List['cooler.Cooler']:\n",
    "    if cpus > 1:\n",
    "        import multiprocessing as mp\n",
    "        pool = mp.Pool(cpus)\n",
    "        my_map = pool.imap_unordered\n",
    "    else:\n",
    "        my_map = map\n",
    "    coolers = pai.get_coolers(cools)\n",
    "    target = pah.get_min_contacts(coolers)\n",
    "    sub_cools = [\"sub_\" + cl for cl in cools]\n",
    "    for i in range(len(cools)):\n",
    "        sample_cooler(\n",
    "            coolers[i],\n",
    "            sub_cools[i],\n",
    "            count=target-1,\n",
    "            exact=False,\n",
    "            map_func=my_map\n",
    "        )\n",
    "    sub_coolers = pai.get_coolers(sub_cools)\n",
    "    # Balance the subsampled versions\n",
    "    if balance:\n",
    "        for clr in sub_coolers:\n",
    "            cooler.balance_cooler(\n",
    "                clr,store=True,\n",
    "                map=my_map,\n",
    "                mad_max=5,\n",
    "                chunksize=10000000,\n",
    "                min_nnz=10,\n",
    "                max_iters=200,\n",
    "                ignore_diags=2\n",
    "            )\n",
    "    if cpus > 1:\n",
    "        pool.close()\n",
    "    return sub_coolers\n",
    "\n",
    "healthy_cool, infected_cool = subsample_cools(['healthy.cool', 'infected.cool'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Only regions with p-values below 10e-3 are selected as potential candidates. When replicates will be available, I will use FDR instead of simple p-value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom1</th>\n",
       "      <th>start1</th>\n",
       "      <th>end1</th>\n",
       "      <th>chrom2</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "      <th>bin1</th>\n",
       "      <th>bin2</th>\n",
       "      <th>diff_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scaffold_4</td>\n",
       "      <td>1496000</td>\n",
       "      <td>1498000</td>\n",
       "      <td>scaffold_4</td>\n",
       "      <td>1496000</td>\n",
       "      <td>1498000</td>\n",
       "      <td>4056.0</td>\n",
       "      <td>4056.0</td>\n",
       "      <td>0.172005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>402000</td>\n",
       "      <td>404000</td>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>402000</td>\n",
       "      <td>404000</td>\n",
       "      <td>201.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>-0.117396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>426000</td>\n",
       "      <td>428000</td>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>426000</td>\n",
       "      <td>428000</td>\n",
       "      <td>213.0</td>\n",
       "      <td>213.0</td>\n",
       "      <td>0.157139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scaffold_5</td>\n",
       "      <td>914000</td>\n",
       "      <td>916000</td>\n",
       "      <td>scaffold_5</td>\n",
       "      <td>914000</td>\n",
       "      <td>916000</td>\n",
       "      <td>4592.0</td>\n",
       "      <td>4592.0</td>\n",
       "      <td>-0.133855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scaffold_8</td>\n",
       "      <td>1298000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>scaffold_8</td>\n",
       "      <td>1298000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>7139.0</td>\n",
       "      <td>7139.0</td>\n",
       "      <td>-0.158142</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom1   start1     end1      chrom2   start2     end2    bin1    bin2  \\\n",
       "0  scaffold_4  1496000  1498000  scaffold_4  1496000  1498000  4056.0  4056.0   \n",
       "1  scaffold_1   402000   404000  scaffold_1   402000   404000   201.0   201.0   \n",
       "2  scaffold_1   426000   428000  scaffold_1   426000   428000   213.0   213.0   \n",
       "3  scaffold_5   914000   916000  scaffold_5   914000   916000  4592.0  4592.0   \n",
       "4  scaffold_8  1298000  1300000  scaffold_8  1298000  1300000  7139.0  7139.0   \n",
       "\n",
       "   diff_score  \n",
       "0    0.172005  \n",
       "1   -0.117396  \n",
       "2    0.157139  \n",
       "3   -0.133855  \n",
       "4   -0.158142  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def filter_patterns(df):\n",
    "    filt = (df.loc[df.diff_score != 0.0, :]\n",
    "      .sort_values(abs(df.diff_score))\n",
    "      .reset_index(drop=True)\n",
    "   )\n",
    "    return filt\n",
    "diff_borders = diff_borders.loc[diff_borders.diff_score != 0.0, :].reset_index(drop=True)\n",
    "diff_loops = diff_loops.loc[diff_loops.diff_score != 0.0, :].reset_index(drop=True)\n",
    "diff_borders.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Keeping 44 borders and 76 loops with detectable change..\n"
     ]
    }
   ],
   "source": [
    "print(f\"Keeping {diff_borders.shape[0]} borders and {diff_loops.shape[0]} loops with detectable change..\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_di(mat, max_pix=10):\n",
    "    \"\"\" Computes directionalitin index according to definition from Dixon et al., 2012\"\"\"\n",
    "    \n",
    "    di = np.zeros(mat.shape[0])\n",
    "    for i in range(mat.shape[0]):\n",
    "        before = np.nanmean(mat[i - max_pix: i, i -  max_pix: i])\n",
    "        after = np.nanmean(mat[i: i + max_pix, i: i +  max_pix])\n",
    "        expected = (before + after ) / 2\n",
    "        sign = ((before - after) / np.abs(before - after))\n",
    "        di[i] = sign * (((after - expected)**2) / expected + ((before - expected)**2) / expected)\n",
    "    di[:max_pix] = 0\n",
    "    di[-max_pix:] = 0\n",
    "    high = np.nanmax(np.abs(di))\n",
    "    di /= high # Scale to 1\n",
    "    return di"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "53d6f6bc1281446eaa19bcad9b840a20",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='region_id', options=(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 648x432 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import set_matplotlib_formats\n",
    "%matplotlib inline\n",
    "set_matplotlib_formats('svg')\n",
    "warnings.filterwarnings('ignore')\n",
    "plt.rcParams['figure.figsize'] = (9.0, 6.0)\n",
    "# Show example regions\n",
    "def plot_regions(healthy_cool, infected_cool, df, region_id, region_size=100000, blur=0):\n",
    "    \n",
    "    def region_lines(ax, s, e):\n",
    "        \"\"\"Add lines to mark region of differential contacts\"\"\"\n",
    "        style = {\"lw\": 0.5, \"alpha\": 0.6, \"c\": \"g\"}\n",
    "        for i in range(3):\n",
    "            ax[0, i].axvline(x=s, **style)\n",
    "            ax[0, i].axvline(x=e, **style)\n",
    "            ax[0, i].axhline(y=s, **style)\n",
    "            ax[0, i].axhline(y=e, **style)\n",
    "            ax[1, i].axvline(x=s, **style)\n",
    "            ax[1, i].axvline(x=e, **style)\n",
    "            \n",
    "    \n",
    "    def nan_gaussian(U, sigma):\n",
    "        \"\"\"Gaussian filter which does not include NAs\"\"\"\n",
    "        import scipy.ndimage as ndi\n",
    "        V=U.copy()\n",
    "        V[np.isnan(U)]=0\n",
    "        VV=ndi.gaussian_filter(V, sigma=sigma)\n",
    "        W=0*U.copy()+1\n",
    "        W[np.isnan(U)]=0\n",
    "        WW=ndi.gaussian_filter(W, sigma=sigma)\n",
    "        Z=VV/WW   \n",
    "        return Z\n",
    "    \n",
    "    # Extract region of interest\n",
    "    region = df.iloc[region_id]\n",
    "    chrom, start, end = region.chrom1, region.start1, region.end1\n",
    "    pos = (start + end) / 2\n",
    "    ucsc_query = f'{chrom}:{int(max(0, pos-region_size))}-{int(pos+region_size)}'\n",
    "    \n",
    "    # Subset matrix\n",
    "    healthy_zoom = healthy_cool.matrix(balance=True).fetch(ucsc_query)\n",
    "    infected_zoom = infected_cool.matrix(balance=True).fetch(ucsc_query)\n",
    "    infected_zoom[np.isnan(infected_zoom)] = 0.0\n",
    "    healthy_zoom[np.isnan(healthy_zoom)] = 0.0\n",
    "    \n",
    "    # Blur ratio to improve readability\n",
    "    if blur > 0:\n",
    "        infected_blur = nan_gaussian(infected_zoom, sigma=blur)\n",
    "        healthy_blur = nan_gaussian(healthy_zoom, sigma=blur)\n",
    "    else:\n",
    "        infected_blur = infected_zoom\n",
    "        healthy_blur = healthy_zoom\n",
    "\n",
    "    log_ratio = np.log2(infected_blur / healthy_blur)\n",
    "    #log_ratio[np.isnan(infected_zoom)] = 0.0\n",
    "    #log_ratio[np.isnan(healthy_zoom)] = 0.0\n",
    "    \n",
    "    # Initialize figure\n",
    "    fig, ax = plt.subplots(2, 3, sharex=True, sharey=False, gridspec_kw={'height_ratios': [5, 1]})\n",
    "    \n",
    "    # Draw lines\n",
    "    region_size = (end - start) // healthy_cool.binsize\n",
    "    mid = healthy_zoom.shape[0] // 2\n",
    "    start_bin, end_bin = mid - region_size, mid + region_size\n",
    "    region_lines(ax, start_bin, end_bin)\n",
    "    \n",
    "    # Make heatmap\n",
    "    plt.suptitle(f\"{chrom}:{start}-{end}\")\n",
    "    ax[0, 0].imshow(np.log2(infected_zoom), cmap=\"Reds\")\n",
    "    ax[0, 0].set_title(\"Infected\")\n",
    "    ax[0, 1].imshow(np.log2(healthy_zoom), cmap=\"Reds\")\n",
    "    ax[0, 1].set_title(\"Uninfected\")\n",
    "    ax[0, 2].imshow(log_ratio, cmap='bwr', vmin=-2, vmax=2)\n",
    "    ax[0, 2].set_title(\"I / U\")\n",
    "    \n",
    "    # Plot DI\n",
    "    n = infected_zoom.shape[0]\n",
    "    infected_di = compute_di(infected_zoom, max_pix=10)\n",
    "    healthy_di = compute_di(healthy_zoom, max_pix=10)\n",
    "    \n",
    "    for i in range(3): ax[1, i].set_ylim(-1, 1, emit=False)\n",
    "    for i in range(3): ax[1, i].axhline(0, lw=0.5, c='black')\n",
    "    ax[1, 0].plot(range(n), infected_di)\n",
    "    ax[1, 1].plot(range(n), healthy_di)\n",
    "    ax[1, 2].plot(range(n), infected_di - healthy_di)\n",
    "\n",
    "\n",
    "\n",
    "from ipywidgets import interact, interactive, fixed, interact_manual\n",
    "import ipywidgets as widgets\n",
    "#show_inter_chr(ot_mat, chrA='chr03', chrB='chr08', var='repeats_prop')\n",
    "region_size_slider = widgets.IntSlider(min=10000, max=500000, step=2000, value=100000)\n",
    "blur_slider = widgets.FloatSlider(min=0, max=4, step=0.1, value=0)\n",
    "pl = interactive(plot_regions, healthy_cool=fixed(healthy_cool),\n",
    "                 infected_cool=fixed(infected_cool),\n",
    "                 df=fixed(diff_borders),\n",
    "                 region_id=range(diff_borders.shape[0]),\n",
    "                 region_size=region_size_slider,\n",
    "                 blur=blur_slider\n",
    "                )\n",
    "#display(pl)\n",
    "display(pl)\n",
    "plt.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "set_matplotlib_formats('svg')\n",
    "warnings.filterwarnings('ignore')\n",
    "plt.rcParams['figure.figsize'] = (9.0, 6.0)\n",
    "# Show example regions\n",
    "\n",
    "\n",
    "#show_inter_chr(ot_mat, chrA='chr03', chrB='chr08', var='repeats_prop')\n",
    "pl = interactive(plot_regions, healthy_cool=fixed(healthy_cool),\n",
    "                 infected_cool=fixed(infected_cool),\n",
    "                 df=fixed(diff_loops),\n",
    "                 region_id=range(diff_loops.shape[0]),\n",
    "                 region_size=region_size_slider,\n",
    "                 blur=blur_slider\n",
    "                )\n",
    "#display(pl)\n",
    "display(pl)\n",
    "plt.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_loops.diff_score.hist()\n",
    "plt.title(\"Differential loop scores (infected - healthy)\")\n",
    "plt.ylabel(\"N loops\")\n",
    "plt.ylabel(\"Differential score\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_borders.diff_score.hist()\n",
    "plt.title(\"Differential border scores (infected - healthy)\")\n",
    "plt.ylabel(\"N borders\")\n",
    "plt.ylabel(\"Differential score\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All genes that overlap regions of differential contacts are then selected. This may not be optimal because genes may be separated from their enhancer element without being affected directly. This can happen if a domain boundary moves into the sequence between gene and enhancer."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The enrichment of GO terms (annotations) within regions with altered chromatin conformation is then tested to check for specific pathways or funcions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Contig</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "      <th>Product</th>\n",
       "      <th>BUSCO</th>\n",
       "      <th>PFAM</th>\n",
       "      <th>InterPro</th>\n",
       "      <th>EggNog</th>\n",
       "      <th>COG</th>\n",
       "      <th>GO Terms</th>\n",
       "      <th>Secreted</th>\n",
       "      <th>Membrane</th>\n",
       "      <th>Protease</th>\n",
       "      <th>CAZyme</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Translation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FUN_000001</td>\n",
       "      <td>CDS</td>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>503</td>\n",
       "      <td>5457</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF00071</td>\n",
       "      <td>IPR001806;IPR020849;IPR027417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0016020;GO:0005525;GO:0003924;GO:0007165</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 (o833-852i)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MATKDFKIAIGGPGGVGKSAIVLQFVTGNYVSEYGAYRKQFALGDN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FUN_000002</td>\n",
       "      <td>CDS</td>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>7855</td>\n",
       "      <td>9815</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MQLMSTQHHKKRTPEDHRPDLPSSARNLLASSPPARSPRSPDGSPH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FUN_000003</td>\n",
       "      <td>CDS</td>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>10018</td>\n",
       "      <td>10602</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF07258</td>\n",
       "      <td>IPR017920</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MWRRPQSVAVGYTGAPKRASLAVGGSGPAVSGRKGDVRFPTLADLR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FUN_000004</td>\n",
       "      <td>CDS</td>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>10748</td>\n",
       "      <td>11427</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 (o90-111i)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MWKQSSRVAVVARSRAVVPSSSGLRRSAGALLPRRAGAETEPEWKL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FUN_000005</td>\n",
       "      <td>CDS</td>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>11709</td>\n",
       "      <td>30538</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF00076;PF00622;PF00632</td>\n",
       "      <td>IPR000504;IPR000569;IPR001870;IPR003877;IPR012...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0004842;GO:0003676;GO:0005515</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2 (o2988-3009i3316-3335o)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGAEEAVHLEGHHSSRSNRPATPQYSAASASGAAPPAPPLRGEAAS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       GeneID Feature      Contig  Start   Stop Strand Name  \\\n",
       "0  FUN_000001     CDS  scaffold_1    503   5457      -  NaN   \n",
       "1  FUN_000002     CDS  scaffold_1   7855   9815      +  NaN   \n",
       "2  FUN_000003     CDS  scaffold_1  10018  10602      -  NaN   \n",
       "3  FUN_000004     CDS  scaffold_1  10748  11427      +  NaN   \n",
       "4  FUN_000005     CDS  scaffold_1  11709  30538      -  NaN   \n",
       "\n",
       "                Product BUSCO                     PFAM  \\\n",
       "0  hypothetical protein   NaN                  PF00071   \n",
       "1  hypothetical protein   NaN                      NaN   \n",
       "2  hypothetical protein   NaN                  PF07258   \n",
       "3  hypothetical protein   NaN                      NaN   \n",
       "4  hypothetical protein   NaN  PF00076;PF00622;PF00632   \n",
       "\n",
       "                                            InterPro  EggNog  COG  \\\n",
       "0                      IPR001806;IPR020849;IPR027417     NaN  NaN   \n",
       "1                                                NaN     NaN  NaN   \n",
       "2                                          IPR017920     NaN  NaN   \n",
       "3                                                NaN     NaN  NaN   \n",
       "4  IPR000504;IPR000569;IPR001870;IPR003877;IPR012...     NaN  NaN   \n",
       "\n",
       "                                      GO Terms  Secreted  \\\n",
       "0  GO:0016020;GO:0005525;GO:0003924;GO:0007165       NaN   \n",
       "1                                          NaN       NaN   \n",
       "2                                          NaN       NaN   \n",
       "3                                          NaN       NaN   \n",
       "4             GO:0004842;GO:0003676;GO:0005515       NaN   \n",
       "\n",
       "                    Membrane Protease CAZyme  Notes  \\\n",
       "0              1 (o833-852i)      NaN    NaN    NaN   \n",
       "1                        NaN      NaN    NaN    NaN   \n",
       "2                        NaN      NaN    NaN    NaN   \n",
       "3               1 (o90-111i)      NaN    NaN    NaN   \n",
       "4  2 (o2988-3009i3316-3335o)      NaN    NaN    NaN   \n",
       "\n",
       "                                         Translation  \n",
       "0  MATKDFKIAIGGPGGVGKSAIVLQFVTGNYVSEYGAYRKQFALGDN...  \n",
       "1  MQLMSTQHHKKRTPEDHRPDLPSSARNLLASSPPARSPRSPDGSPH...  \n",
       "2  MWRRPQSVAVGYTGAPKRASLAVGGSGPAVSGRKGDVRFPTLADLR...  \n",
       "3  MWKQSSRVAVVARSRAVVPSSSGLRRSAGALLPRRAGAETEPEWKL...  \n",
       "4  MGAEEAVHLEGHHSSRSNRPATPQYSAASASGAAPPAPPLRGEAAS...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "genes_pr = pr.PyRanges(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_borders_pr = pr.PyRanges(\n",
    "    diff_borders\n",
    "    .loc[:, [\"chrom1\", \"start1\", \"end1\", \"diff_score\"]]\n",
    "    .rename(columns={\"chrom1\": \"Chromosome\", \"start1\": \"Start\", \"end1\": \"End\", \"diff_score\": \"Score\"})\n",
    ")\n",
    "diff_loops_pr = pr.PyRanges(\n",
    "    pd.concat([\n",
    "        (diff_loops\n",
    "            .loc[:, [\"chrom1\", \"start1\", \"end1\", \"diff_score\"]]\n",
    "            .rename(columns={\"chrom1\": \"Chromosome\", \"start1\": \"Start\", \"end1\": \"End\"})),\n",
    "        (diff_loops\n",
    "        .loc[:, [\"chrom2\", \"start2\", \"end2\", \"diff_score\"]]\n",
    "        .rename(columns={\"chrom2\": \"Chromosome\", \"start2\": \"Start\", \"end2\": \"End\"}))\n",
    "    ])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Overlap with annotations\n",
    "diff_borders_genes = genes_pr.overlap(diff_borders_pr).df\n",
    "diff_loops_genes = genes_pr.overlap(diff_loops_pr).df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract known proteins, excluding tRNA genes for visual inspection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_stats(df, name):\n",
    "    is_known = df.Product[df.Product != 'hypothetical protein']\n",
    "    known = len(is_known[is_known])\n",
    "    is_trna = df.Product.str.contains('tRNA-')\n",
    "    trna = len(is_trna[is_trna])\n",
    "    print(\n",
    "        f\"There are {df.shape[0]} {name}, of which {known} have known functions, \"\n",
    "        f\"including {trna} tRNAs.\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "known_loops_genes = diff_loops_genes.loc[(~diff_loops_genes.Product.str.contains('tRNA-')) & (diff_loops_genes.Product != 'hypothetical protein'), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "known_borders_genes = diff_borders_genes.loc[(~diff_borders_genes.Product.str.contains('tRNA-')) & (diff_borders_genes.Product != 'hypothetical protein'), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 14354 genes in the genome, of which 735 have known functions, including 387 tRNAs.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "gene_stats(genes, \"genes in the genome\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 1704 infection dependent loop genes, of which 55 have known functions, including 25 tRNAs.\n"
     ]
    }
   ],
   "source": [
    "gene_stats(diff_loops_genes, 'infection dependent loop genes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 3063 infection dependent border genes, of which 125 have known functions, including 63 tRNAs.\n"
     ]
    }
   ],
   "source": [
    "gene_stats(diff_borders_genes, 'infection dependent border genes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|      | GeneID     | Feature   | Chromosome   |   Start |     End | Strand   | Name        | Product                                                                           |\n",
      "|-----:|:-----------|:----------|:-------------|--------:|--------:|:---------|:------------|:----------------------------------------------------------------------------------|\n",
      "|   15 | FUN_000207 | CDS       | scaffold_1   |  616228 |  618206 | +        | GSK3B       | Glycogen synthase kinase-3 beta                                                   |\n",
      "|  126 | FUN_000336 | CDS       | scaffold_1   | 1007079 | 1008861 | -        | FTSJ1       | Putative tRNA (cytidine(32)/guanosine(34)-2'-O)-methyltransferase                 |\n",
      "|  136 | FUN_000422 | CDS       | scaffold_1   | 1323317 | 1324222 | -        | AP1S2       | AP-1 complex subunit sigma-2                                                      |\n",
      "|  151 | FUN_000568 | CDS       | scaffold_1   | 1720411 | 1721486 | -        | RAB2A_1     | Ras- protein Rab-2A                                                               |\n",
      "|  354 | FUN_009244 | CDS       | scaffold_3   |  264055 |  264662 | +        | RAS1_6      | Ras GTPase                                                                        |\n",
      "|  424 | FUN_009323 | CDS       | scaffold_3   |  520976 |  521806 | -        | LSM5        | RNA-binding protein lsm5                                                          |\n",
      "|  441 | FUN_009427 | CDS       | scaffold_3   |  867864 |  869863 | -        | ARP3        | Arp2/3 complex subunit, actin nucleation center                                   |\n",
      "|  454 | FUN_009497 | CDS       | scaffold_3   | 1101929 | 1103180 | -        | RAB7A_2     | Ras- protein Rab-7A                                                               |\n",
      "|  509 | FUN_011140 | CDS       | scaffold_4   |  784954 |  786643 | +        | VATD1       | V-type proton ATPase subunit d                                                    |\n",
      "|  539 | FUN_010950 | CDS       | scaffold_4   |  149639 |  153140 | -        | DHX8        | ATP-dependent RNA helicase dhx8                                                   |\n",
      "|  557 | FUN_011105 | CDS       | scaffold_4   |  649679 |  651409 | -        | SDH2        | succinate dehydrogenase complex, subunit B                                        |\n",
      "|  619 | FUN_011838 | CDS       | scaffold_5   |  837818 |  838547 | +        | PEX4        | E2 ubiquitin-protein ligase peroxin 4                                             |\n",
      "|  706 | FUN_011961 | CDS       | scaffold_5   | 1213249 | 1215274 | -        | NMT2        | Glycylpeptide N-tetradecanoyltransferase 2                                        |\n",
      "|  711 | FUN_012008 | CDS       | scaffold_5   | 1335191 | 1335823 | -        | RAS85D      | Ras-like protein 1                                                                |\n",
      "|  714 | FUN_012024 | CDS       | scaffold_5   | 1381484 | 1385765 | -        | DHX38       | Pre-mRNA-splicing factor ATP-dependent RNA helicase PRP16                         |\n",
      "|  716 | FUN_012041 | CDS       | scaffold_5   | 1437368 | 1439158 | -        | GPI8        | glycosylphosphatidylinositol anchor biosynthesis                                  |\n",
      "|  748 | FUN_012449 | CDS       | scaffold_6   |  871983 |  873665 | +        | ARO2        | bifunctional chorismate synthase/riboflavin reductase [NAD(P)H] aro2              |\n",
      "|  793 | FUN_012349 | CDS       | scaffold_6   |  524360 |  525323 | -        | EIF6        | eukaryotic translation initiation factor 6                                        |\n",
      "|  836 | FUN_012767 | CDS       | scaffold_7   |   91797 |   93326 | +        | RPL3        | 60S ribosomal protein L3                                                          |\n",
      "| 1051 | FUN_014145 | CDS       | scaffold_9   |  916975 |  918054 | +        | ERD2        | endoplasmic reticulum retention protein                                           |\n",
      "| 1135 | FUN_000899 | CDS       | scaffold_10  |  265031 |  265622 | +        | RAS1_1      | Ras GTPase                                                                        |\n",
      "| 1141 | FUN_000952 | CDS       | scaffold_10  |  422617 |  424435 | +        | DMC1B       | Meiotic recombination protein DMC1 B                                              |\n",
      "| 1161 | FUN_001191 | CDS       | scaffold_10  | 1208650 | 1209025 | +        | RPL35       | 60S ribosomal protein L35                                                         |\n",
      "| 1187 | FUN_000940 | CDS       | scaffold_10  |  363740 |  364421 | -        | CB5LP       | Cytochrome B5-like protein                                                        |\n",
      "| 1361 | FUN_002146 | CDS       | scaffold_12  | 1190354 | 1192506 | +        | IDH2        | NAD-dependent isocitrate dehydrogenase                                            |\n",
      "| 1590 | FUN_002906 | CDS       | scaffold_14  |  690545 |  695191 | -        | ELP3_2      | Elongator subunit                                                                 |\n",
      "| 1617 | FUN_003063 | CDS       | scaffold_14  | 1174720 | 1175227 | -        | RPL32       | 60S ribosomal protein L32                                                         |\n",
      "| 1628 | FUN_003201 | CDS       | scaffold_15  |  247705 |  250086 | +        | CCT6A       | T-complex protein 1 subunit zeta                                                  |\n",
      "| 1714 | FUN_003564 | CDS       | scaffold_16  |   26872 |   27739 | +        | AP3S1       | AP-3 complex subunit sigma-1                                                      |\n",
      "| 1718 | FUN_003606 | CDS       | scaffold_16  |  155696 |  156166 | +        | RPS15AA     | 40S ribosomal protein S15a-1                                                      |\n",
      "| 1729 | FUN_003657 | CDS       | scaffold_16  |  333249 |  337799 | +        | CCT3        | T-complex protein 1 subunit gamma                                                 |\n",
      "| 1742 | FUN_003800 | CDS       | scaffold_16  |  839976 |  841430 | +        | IMPA4_1     | Importin subunit alpha-4                                                          |\n",
      "| 1795 | FUN_004068 | CDS       | scaffold_17  |  499806 |  501727 | +        | NOP56       | snoRNP complex protein nop56                                                      |\n",
      "| 1803 | FUN_004182 | CDS       | scaffold_17  |  830692 |  831560 | +        | ARF79F      | ADP-ribosylation factor 1                                                         |\n",
      "| 1830 | FUN_003973 | CDS       | scaffold_17  |  171845 |  174000 | -        | LYS4        | mitochondrial Homoaconitase                                                       |\n",
      "| 1880 | FUN_004395 | CDS       | scaffold_18  |  120192 |  122021 | +        | PHO2B_1     | Protein phosphatase 2A catalytic subunit B                                        |\n",
      "| 1931 | FUN_004402 | CDS       | scaffold_18  |  140905 |  143183 | -        | DPH1        | Diphthamide biosynthesis protein 1                                                |\n",
      "| 2003 | FUN_005115 | CDS       | scaffold_19  | 1044564 | 1046676 | +        | DPAGT1      | UDP-N-acetylglucosamine--dolichyl-phosphate N-acetylglucosaminephosphotransferase |\n",
      "| 2091 | FUN_006138 | CDS       | scaffold_20  |  975491 |  975932 | +        | TRAPPC4     | Trafficking protein particle complex subunit 4                                    |\n",
      "| 2095 | FUN_006157 | CDS       | scaffold_20  | 1029253 | 1029712 | +        | PROFILIN2_2 | Profilin-2                                                                        |\n",
      "| 2176 | FUN_006373 | CDS       | scaffold_21  |  584422 |  586051 | +        | GPN1        | GPN-loop GTPase 1                                                                 |\n",
      "| 2252 | FUN_006637 | CDS       | scaffold_22  |  186430 |  187920 | +        | RAS1_2      | Ras GTPase                                                                        |\n",
      "| 2266 | FUN_006735 | CDS       | scaffold_22  |  474393 |  475564 | +        | RAB21       | Ras- protein Rab-21                                                               |\n",
      "| 2351 | FUN_007142 | CDS       | scaffold_23  |  626965 |  629628 | +        | MIS3        | Ribosomal RNA assembly protein mis3                                               |\n",
      "| 2361 | FUN_007196 | CDS       | scaffold_23  |  824986 |  826184 | +        | VPS29       | Vacuolar protein sorting-associated protein 29                                    |\n",
      "| 2364 | FUN_007218 | CDS       | scaffold_23  |  878784 |  880020 | +        | PAA1        | Polyamine N-acetyltransferase 1                                                   |\n",
      "| 2376 | FUN_006946 | CDS       | scaffold_23  |   48835 |   49463 | -        | ACAD10      | Acyl-CoA dehydrogenase member 10                                                  |\n",
      "| 2377 | FUN_006947 | CDS       | scaffold_23  |   49491 |   50883 | -        | IBR3        | Putative acyl-CoA dehydrogenase ibr3                                              |\n",
      "| 2405 | FUN_007118 | CDS       | scaffold_23  |  549793 |  551193 | -        | GPN3        | GPN-loop GTPase 3                                                                 |\n",
      "| 2435 | FUN_007365 | CDS       | scaffold_24  |  202717 |  204082 | +        | RBBP4       | Histone-binding protein rbbp4                                                     |\n",
      "| 2554 | FUN_007718 | CDS       | scaffold_25  |  242588 |  243787 | -        | ACTIN1_7    | Actin-1                                                                           |\n",
      "| 2566 | FUN_007812 | CDS       | scaffold_25  |  525234 |  527894 | -        | RNR1        | ribonucleotide-diphosphate reductase subunit rnr1                                 |\n",
      "| 2567 | FUN_007819 | CDS       | scaffold_25  |  544573 |  548387 | -        | NADSYN1     | Glutamine-dependent NAD(+) synthetase                                             |\n",
      "| 2603 | FUN_008115 | CDS       | scaffold_26  |  371789 |  372252 | +        | RAS1_5      | Ras GTPase                                                                        |\n",
      "| 2676 | FUN_008504 | CDS       | scaffold_27  |  466106 |  469450 | +        | ACO2_2      | aconitate hydratase                                                               |\n",
      "| 2733 | FUN_008717 | CDS       | scaffold_28  |  119319 |  121052 | +        | TSTA3       | GDP-L-fucose synthase                                                             |\n",
      "| 2923 | FUN_010046 | CDS       | scaffold_31  |   93584 |   94107 | -        | RPS21       | 40S ribosomal protein S21                                                         |\n",
      "| 2935 | FUN_010168 | CDS       | scaffold_31  |  452400 |  454013 | -        | CLPB1_2     | Chaperone protein ClpB1                                                           |\n",
      "| 2956 | FUN_010329 | CDS       | scaffold_32  |   91131 |   93392 | +        | MAP2B       | Methionine aminopeptidase 2B                                                      |\n",
      "| 3001 | FUN_010535 | CDS       | scaffold_33  |   55085 |   55460 | +        | BRK1        | Protein BRICK1                                                                    |\n",
      "| 3002 | FUN_010540 | CDS       | scaffold_33  |   84147 |   85203 | +        | ACTIN1_8    | Actin-1                                                                           |\n",
      "| 3030 | FUN_010643 | CDS       | scaffold_34  |   63041 |   63825 | +        | MLH1        | DNA mismatch repair protein                                                       |\n"
     ]
    }
   ],
   "source": [
    "print(known_borders_genes.loc[:, [\"GeneID\", \"Feature\", \"Chromosome\", \"Start\", \"End\", \"Strand\", \"Name\", \"Product\",]].to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute GO enrichment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_all_genes = {x: GeneID2nt_mus[x] for x in genes.entrezgene_id if x in GeneID2nt_mus.keys()}\n",
    "exp_diff_genes = {x: GeneID2nt_mus[x] for x in genes_diff.entrezgene_id if x in GeneID2nt_mus.keys()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_diff_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build GO enrichment analysis object\n",
    "\n",
    "goeaobj = GOEnrichmentStudyNS(\n",
    "        exp_all_genes.keys(), # List of mouse protein-coding genes\n",
    "        ns2assoc, # geneid/GO associations\n",
    "        obodag, # Ontologies\n",
    "        propagate_counts = False,\n",
    "        alpha = 0.05, # default significance cut-off\n",
    "        methods = ['fdr_bh']) # defult multipletest correction method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run gene ontology enrichment analysis\n",
    "# 'p_' means \"pvalue\". 'fdr_bh' is the multipletest method we are currently using.\n",
    "geneids_study = exp_diff_genes.keys()\n",
    "goea_results_all = goeaobj.run_study(geneids_study)\n",
    "goea_results_sig = [r for r in goea_results_all if r.p_fdr_bh < 0.05]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When running the analysis with a stringent threshold of pvalue < 10-4, I get no enriched biological process or molecular function, but only cellular component \"Nucleosome\" with genes Tnp2, Prm1, Prm2 and Prm3.\n",
    "\n",
    "When running the analysis with less stringent filter pvalue < 10-3, I get enrichment for biological processes keratinization, keratinocyte differentiation and peptide cross linking. This is due to a group of several Sprr genes in the region at chr3:92Mb-94Mb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib notebook\n",
    "# Save and visualize results\n",
    "gene2sym = { k: v.Symbol for k, v in exp_diff_genes.items()}\n",
    "goeaobj.wr_txt(out+\"go_enrichment.txt\", goea_results_sig)\n",
    "plot_results(out+\"mouse_salmonella_GO_enriched_{NS}.png\",\n",
    "             goea_results_sig,\n",
    "             id2symbol=gene2sym,\n",
    "             study_items=6)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "img=mpimg.imread(out+\"mouse_salmonella_GO_enriched_BP.png\")\n",
    "plt.figure()\n",
    "plt.imshow(img)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "img=mpimg.imread(out+\"mouse_salmonella_GO_enriched_CC.png\")\n",
    "plt.figure()\n",
    "plt.imshow(img)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "img=mpimg.imread(out+\"mouse_salmonella_GO_enriched_MF.png\")\n",
    "plt.figure()\n",
    "plt.imshow(img)\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
